project-lisa/source-code/client/src/api.ts

```diff
@@ -1,24 +1,261 @@
+	push: (args: { dir: string }) => Promise<Awaited<ReturnType<typeof raw.push>> | undefined>
+	pull: (args: { dir: string, author: any, fastForward: boolean, singleBranch: true }) => any
+	add: (args: { dir: string, filepath: string }) =>  Promise<Awaited<ReturnType<typeof raw.add>>>
+	listRemotes: () => Promise<Awaited<ReturnType<typeof raw.listRemotes>> | undefined>
+	log: (args: { dir: string, since: any}) =>  Promise<Awaited<ReturnType<typeof raw.log>>>
+	statusMatrix: (args: { dir:string, filter: any }) => Promise<Awaited<ReturnType<typeof raw.statusMatrix>>>
+	status: (args: { dir: string, filepath: string }) => Promise<Awaited<ReturnType<typeof raw.status>>>
+	mergeUpstream: (args: { branch: string }) => ReturnType<typeof github.request<"POST /repos/{owner}/{repo}/merge-upstream">>
+	isCollaborator: (args: { username: string }) => Promise<boolean>
+	getOrigin: () => Promise<string>
+	getCurrentBranch: () => Promise<string | undefined>
+	getMeta: () => Promise<{
+		name: string,
+		isPrivate: boolean,
+		isFork: boolean,
+		owner: { name?: string, email?: string },
+		parent:  {
+			url: string,
+			fullName: string
+		} | undefined
+	}>
+
+	// TODO: implement these before publishing api, but not used in badge or editor
+	// currentBranch: () => unknown
+	// changeBranch: () => unknown
+}
+
+export function load ({ url, fs, corsProxy }: { url: string, fs?: NodeishFilesystem, path?: string, corsProxy?: string, auth?: unknown }): Repository {
+	const rawFs = fs ? fs : createNodeishMemoryFs()
+
+	// parse url in the format of github.com/inlang/example and split it to host, owner and repo
+	const [host, owner, repoName] = [...url.split("/")]
+
+	// check if all 3 parts are present, if not, return an error
+	if (!host || !owner || !repoName) {
+		throw new Error(
+			`Invalid url format for '${url}' for cloning repository, please use the format of github.com/inlang/example.`,
+		)
+	}
+
+	const normalizedUrl = `https://${host}/${owner}/${repoName}`
+
+	let pending: Promise<void> | undefined = raw.clone({
+		fs: wrap(rawFs, 'clone'),
+		http,
+		dir: "/",
+		corsProxy,
+		url: normalizedUrl,
+		singleBranch: true,
+		depth: 1,
+		noTags: true,
+	}).finally(() => {
+		pending = undefined
+	}).catch((err: any) => {
+		console.error('error cloning the repository', err)
+	})
+
+	// delay all fs and repo operations until the repo clone and checkout have finished, this is preparation for the lazy feature
+	function delayedAction ({ execute }: { execute: () => any }) {
+		if (pending) {
+			return pending.then(execute)
+		}
+
+		return execute()
+	}
+
+	return {
+		fs: wrap(rawFs, 'app', delayedAction),
+
+		/**
+		 * Gets the git origin url of the current repository.
+		 *
+		 * @returns The git origin url or undefined if it could not be found.
+		 */
+		async listRemotes () {
+			try {
+				const wrappedFS = wrap(rawFs, 'listRemotes', delayedAction)
+
+				const remotes = await raw.listRemotes({
+					fs: wrappedFS,
+					dir: await raw.findRoot({
```

Comment:
Seems like we can extract `findRoot` to the scope of a repo. The concept of changing working directories does not exist in lisa/client.

Reply:
No need to call findRoot for every listRemotes call. There is always only one root, the root of the repo.